### PR TITLE
ASoC: SOF: Intel: Use the old and simpler IMR boot method with IPC3

### PR DESCRIPTION
**Pull request by ujfalusi**
The new sequence started to cause random and rare IMR boot failures with
IPC3 firmware which manifests something like this:

IMR restore supported, booting from IMR directly
DSP core(s) enabled? 0 : core_mask 3
FW Poll Status: reg[0x4]=0x3030303 successful
FW Poll Status: reg[0x4]=0x3030300 successful
unstall/run core: core_mask = 3
DSP core(s) enabled? 1 : core_mask 3
FW Poll Status: reg[0x4]=0x3030000 successful
unstall/run core: core_mask = 1
DSP core(s) enabled? 1 : core_mask 1
FW Poll Status: reg[0xd4]=0x80000000 successful
FW Poll Status: reg[0x4]=0x3030202 successful
FW Poll Status: reg[0x4]=0x1010202 successful
DSP core(s) enabled? 0 : core_mask 2
no reply expected, received 0x0, will be ignored
no reply expected, received 0x0, will be ignored
FW Poll Status: reg[0x80000]=0x5 successful
no reply expected, received 0x0, will be ignored
no reply expected, received 0x0, will be ignored
...

and the FW_READY never arrives.
The reproduction rate is really low, but never the less it has been
observed multiple times.

The likely reason for this is the ROM message sending and signaling around
it which confuses the firmware and the hardware when IPC3 based firmware is
used.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

### CODE REVIEW

**Comment by ujfalusi** [comment]
Recently we started to see failures with IMR booting on few platforms: #3586, it looks like we did not had similar ones before the IMR boot sequence rework was applied (8th of April). The reproduction rate is low, but there is definitely something going on which points to #3558.

File: sound/soc/sof/intel/hda-loader.c [hunk 1]
```diff
@@ -369,10 +369,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
...
+	}
+
+	/*
+	 * For IPC3 the IMR boot sequence is much simpler and straight forward,
+	 * no ROM message is needed to be sent for example
+	 */
+
+	/* step 1: power up & unstall/run the cores to run the firmware */
+	ret = hda_dsp_enable_core(sdev, chip->init_core_mask);
+	if (ret < 0) {
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-loader.c:397]: Will be `if (ret)`

File: sound/soc/sof/intel/hda-loader.c [hunk 2]
```diff
@@ -369,10 +369,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
...
+	unsigned long mask;
+	int i, ret;
+
+	if (sdev->pdata->ipc_type == SOF_INTEL_IPC4) {
+		/*
+		 * For IPC4 the IMR boot flow follows the normal boot with subtle
+		 * differences
+		 */
+		ret = cl_dsp_init(sdev, 0, true);
+		if (ret >= 0)
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-loader.c:384]: Will be `if (!ret)`

**Comment by ujfalusi** [comment]
it might be that the boot sequence in cl_dsp_init() have a bug, not needed step which cause the issue:
https://github.com/thesofproject/linux/blob/035fa5dbf5ed398d87e8df22a8dc9644a34a4ed4/sound/soc/sof/intel/hda-loader.c#L155-L159

None of the documentation I have checked so far mentioned anything like ack/DONE for the ROM boot config message (ROM_CONTROL), all it is written is: place the pseudo IPC message for ROM config into IPC register, unstall core0 and this will take care of the sending of message to ROM, and it is optional.

I have tested it locally on tgl-h and it works fine (removing the said lines).

**Comment by ujfalusi** [comment]
To test the DONE bit handling removal on wider range of machines: #3623

**Comment by ranj063** [comment]
> it might be that the boot sequence in cl_dsp_init() have a bug, not needed step which cause the issue:
> 
> https://github.com/thesofproject/linux/blob/035fa5dbf5ed398d87e8df22a8dc9644a34a4ed4/sound/soc/sof/intel/hda-loader.c#L155-L159
> 
> None of the documentation I have checked so far mentioned anything like ack/DONE for the ROM boot config message (ROM_CONTROL), all it is written is: place the pseudo IPC message for ROM config into IPC register, unstall core0 and this will take care of the sending of message to ROM, and it is optional.
> 
> I have tested it locally on tgl-h and it works fine (removing the said lines).

@ujfalusi this step was added to prevent an unsolicited response rfom the DSP after FW load. It did not exist in the original sequence.

commit 37a079415456b34520454828d582b5bbdad0473f
Author: Keyon Jie <yang.jie@linux.intel.com>
Date:   Sat Nov 9 05:21:00 2019 +0800

    ASoC: SOF: Intel: hda-loader: clear the IPC ack bit after FW_PURGE done

**Comment by ranj063** [comment]
> it might be that the boot sequence in cl_dsp_init() have a bug, not needed step which cause the issue:
> 
> https://github.com/thesofproject/linux/blob/035fa5dbf5ed398d87e8df22a8dc9644a34a4ed4/sound/soc/sof/intel/hda-loader.c#L155-L159
> 
> None of the documentation I have checked so far mentioned anything like ack/DONE for the ROM boot config message (ROM_CONTROL), all it is written is: place the pseudo IPC message for ROM config into IPC register, unstall core0 and this will take care of the sending of message to ROM, and it is optional.
> 
> I have tested it locally on tgl-h and it works fine (removing the said lines).

Yes, it does work without those lines but the problem is that when we receive that unsolicited response from theFW after boot, the driver will complain about it even though it might actually be harmless. This leads to any issues during boot being pointed to the audio driver because of the warning.

**Comment by ujfalusi** [comment]
Can we wait for few more days to make sure that the #3586 is still happening? I have been scanning the daily reports and have not seen it popping up. It might have been caused something else and by co incident it started to pop up after #3558 got merged.
I don't hold my breath, but it is a possibility.
This PR can quickly applied as a remedy if it is still happening.

**Comment by ujfalusi** [comment]
It looks like IPC4 also shows issues with the full boot sequence, see #3636

**Pull request closed without merging**